### PR TITLE
Fix for Issue #79

### PR DESCRIPTION
**Pull request by yihchun**
Allows for per-SDR PPM information.

- Legacy support for stratux:978 and stratux:1090 is preserved
- SDRs without a specified PPM get the globalSettings PPM as configured on the Settings page
- Because the serial number field is length-constrained, the preferred format with PPM is stx:freq:ppm, such as stx:1090:20. Per peepsnet's question below, negative PPMs are supported; example: stx:1090:-20
- The PPM parser and the frequency designator are independent entities. If you don't need to specify a frequency, pick any number other than 978 and 1090 (0 recommended), and Stratux will assign frequencies. (Per peepsnet's second question below). Example: stx:0:-20

### CODE REVIEW

**Comment by peepsnet** [comment]
Will this work for negative ppm vals??  stx:978:-25

**Comment by yihchun** [comment]
Yes, the regexps include a -? to capture possibly negative numbers.

**Comment by peepsnet** [comment]
Great! Im sold!!

**Comment by peepsnet** [comment]
Another question. There was a post made to address a SDR failure in flight. If the 978 SDR failed in flight and your other SDR's serial was stratux:1090 then you would never be able to get weather after a reboot because the working SDR would tune to 1090.  The Solution was to set only one SDR serial to stratux:978 and the other to something else(stratux:atl suggested) This would allow the this SDR to not be forced to 1090 if detected first. 

Does your code allow for this?  
SDR 1: stx:978:-24
SDR 2: stx:alt:-27

Would the code set SDR 2 to 978 if there is only one SDR detected but 978 & 1090 turned on in the settings??

Does this make sense??

**Comment by yihchun** [comment]
Doesn't allow for your format, but stx:0:-27 should work.

**Comment by peepsnet** [comment]
ok. just something to allow the functionality.  So if there is no 978 SDR found but 978 is wanted(turned on in settings) then the remaining SDR will be set to 978 if the serial is not 1090.

**Comment by yihchun** [comment]
Right. In fact, if you don't have frequency-specific antennas, you can set both of your radios to stx:0:-24 and stx:0:-27 and Stratux will always decide for you.

**Comment by peepsnet** [comment]
What freq is set first?  is it 978 by default??

**Comment by yihchun** [comment]
If you only have one, it'll shut down the 1090 to switch it to 978, unless 1090 is explicitly set in serial.
"When count is one, favor UAT in the case where the user has enabled both UAT and ES via the web interface."

**Comment by peepsnet** [comment]
That is the Correct logic!  I believe that is a favored logical path for a failure. You would want weather more then traffic...  Great job on the code. That is the idea from the piaware software, correct? I believe I saw this on an internet search. Glad it can be added here!!

**Comment by peepsnet** [comment]
I think the next step for this is to have a check box in the settings webui which allows the user to determine if they want to override the 1090 SDR if the 978 fails. This would allow the user to set the serials to match the SDR's and not lose the prioritization

**Comment by yihchun** [comment]
The code doesn't borrow from piaware or anything; it's just a simple regexp matching.

As for webui / default behavior changes, that's out of the scope of this pull request, which is limited to changing the format of the serial number, and integrating the PPM therein. It's also beyond the scope of Issue #79 .

**Comment by peepsnet** [comment]
I know that is beyond this pull... But to allow anybody access to this functionality it should be added to settings and allowed to choose priority. But in order to get the PPM offset I guess one would need some knowledge. Enough to not need a web interface. 

But maybe someday it can all be in the webUI... Even the kal functionality...

File: main/sdr.go [hunk 1]
```diff
@@ -365,7 +384,9 @@ func sdrWatcher() {
 				if err != nil {
 					serial = ""
 				}
-				if strings.Compare(serial, "stratux:1090") != 0 {
+
+				r, _ := regexp.Compile("str?a?t?u?x:1090")
```
Comment (jpoirier) [inline main/sdr.go]: The regexp string isn't changing so no reason to compile it with each pass, just do it once on entry in to the function prior to dropping in to the for loop. Why no error check/handling?

File: main/sdr.go [hunk 2]
```diff
@@ -407,7 +428,8 @@ func sdrWatcher() {
 				if err != nil {
 					serial = ""
 				}
-				if strings.Compare(serial, "stratux:978") != 0 {
+				r, _ := regexp.Compile("str?a?t?u?x:978")
```
Comment (jpoirier) [inline main/sdr.go]: ...same comment as for r, _ := regexp.Compile("str?a?t?u?x:1090")

File: main/sdr.go [hunk 3]
```diff
@@ -124,12 +126,29 @@ func (u *UAT) read() {
 	}
 }
 
+func getPPM(serial string) int {
+	r, _ := regexp.Compile("str?a?t?u?x:\\d+:?(-?\\d*)")
```
Comment (jpoirier) [inline main/sdr.go]: Why no error check/handling?

**Comment by yihchun** [comment]
Thanks for the comments. I didn't error check the regexp compiles because I figured the regexps are static, so they should compile OK. I've put in error checking code now, with a fallback to string.Compare.

**Comment by jpoirier** [comment]
LGTM but the merge is Chris' call

**Comment by cyoung** [comment]
Thanks @yihchun, and thanks @jpoirier for reviewing.

**Comment by cyoung** [comment]
@yihchun - need to set up your github client so that your commits aren't showing up as "root"

**Comment by yihchun** [comment]
Will do, sorry about that.

**Comment by peepsnet** [comment]
Is the PPM setting on the Settings page still used at all? And if so when?
And did you make any log entries for setting the ppm?
And what was the format? stx:xxx:ppm or str:xxx:ppm? The ereg starts str?a?

It apprears the code behind the ppm field in settings is broken. If I delete the text it returns after leaving the page and returning.

**Comment by yihchun** [comment]
PPM setting is used when no explicit per-dongle PPM is set.

The log entries for UAT PPM remains unchanged; there was no prior log message for ES PPM, and I did not add one.

The format is stx:number:ppm. ? means that the previous element is optional.

I didn't touch the settings field.

**Pull request merged**